rfcs/0009-abstraction-layer-for-mobile-error-tracking.md

```diff
@@ -0,0 +1,293 @@
+
+  const setUserId = (level) =>
+    sentry.configureScope((scope) => scope.setLevel(level));
+
+  // You can set the severity of an event to one of five values: fatal, error, warning, info, and debug. error is the default, fatal is the most severe, and debug is the least severe.
+
+  const setLevel = (id) =>
+    sentry.configureScope((scope) => scope.setUser({ id }));
+
+  const enable = (enabled) => {
+    isTrackingEnabled = enabled;
+
+    return Promise.resolve()
+  };
+
+  return {
+    captureError: createTrackingFn(captureError),
+    addBreadcrumb: createTrackingFn(addBreadcrumb),
+    setUserId: createTrackingFn(setUserId),
+    setLevel: createTrackingFn(setLevel),
+    enable,
+  };
+};
+
+export default createErrorTracking;
+```
+
+## Implementation for Firebase Crashlytics
+
+```js
+const createErrorTracking = (crashlytics, { enabled = false } = {}) => {
+  const enable = (enabled) => {
+    crashlytics.setCrashlyticsCollectionEnabled(enabled);
+  };
+
+  await enable(enabled);
+
+  const captureError = crashlytics.recordError;
+
+  const setUserId = crashlytics.setUserId;
+
+  const setLevel = (level) => crashlytics.setAttributes({ level });
+
+  // level should be: fatal, error, warning, info, or debug.
+  const addBreadcrumb = ({ scope, message, level = "info" }) =>
+    crashlytics.setAttributes({ category: scope, message, level });
+
+  return {
+    captureError,
+    addBreadcrumb,
+    setLevel,
+    setUserId,
+    enable,
+  };
+};
+
+export default createErrorTracking;
+```
+
+## Setup
+
+```js
+// src/shared/react-error-tracking.js
+
+import React, { createContext, useContext } from "react";
+
+const ErrorTrackingContext = createContext();
+
+export const ErrorTrackingProvider = ({ children }) => {
+  return (
+    <ErrorTrackingContext.Provider>{children}</ErrorTrackingContext.Provider>
+  );
+};
+
+export const useErrorTracking = () => {
+  const context = useContext(ErrorTrackingContext);
+
+  if (!context) {
+    throw new Error(
+      "useErrorTracking should be used within an ErrorTrackingProvider"
+    );
+  }
+
+  return context;
+};
+```
+
+```js
+// src/app/App.js
+import * as Sentry from "@sentry/react-native";
+import createErrorTracking from "@moxy/react-native-sentry";
+import { ErrorTrackingProvider } from "../../shared/react-error-tracking";
+
+Sentry.init({
+  dsn: // Sentry's DSN,
+});
+
+const errorTracking = createErrorTracking(Sentry, { enabled: true });
+
+const App = () => (
+  <ErrorTrackingProvider value={errorTracking}>
+    {/* ... */}
+  </ErrorTrackingProvider>
+);
+```
+
+## Usage examples
+
+Suppose there is a `Button.js` component:
+
+```js
+import React, { useCallback } from "react";
+import { TouchableOpacity, Text } from "react-native";
+import { useErrorTracking } from "../react-error-tracking";
+
+const aFunctionThatMightFail = () => {
+  throw new Error("Failed");
+};
+
+const Button = () => {
+  const errorTracking = useErrorTracking();
+
+  const onPress = useCallback(() => {
```

Comment:
Just a suggestion, in these cases we could create a hook that uses `useCallback` and performs the try/catch with error tracking. This way we would only use a hook like we would use `useCallback` and the try/catch would be "automatic"

Reply:
Thanks for your suggestion @ruipneves ! Didn't think about this possibility to be honest. I will definitely keep that in mind once I start working on the implementation